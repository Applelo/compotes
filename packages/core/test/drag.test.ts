import { afterAll, beforeAll, describe, expect, it } from 'vitest'
import type { Browser, Page } from 'playwright'
import { chromium } from 'playwright'

let browser: Browser

beforeAll(async () => {
  browser = await chromium.launch()
})

afterAll(async () => {
  browser.close()
})

async function dragBox(page: Page) {
  const dragEl = page.locator('.c-drag').first()
  const lastDragEl = dragEl.locator('p:last-child').last()

  expect(dragEl).not.toBeNull()
  expect(lastDragEl).not.toBeNull()
  if (!dragEl || !lastDragEl)
    return {}

  const dragElBoundingBox = await dragEl.boundingBox()
  const previousLastDragElBoundingBox = await lastDragEl.boundingBox()

  expect(dragElBoundingBox).not.toBeNull()
  expect(previousLastDragElBoundingBox).not.toBeNull()
  if (!dragElBoundingBox || !previousLastDragElBoundingBox)
    return {}

  await page.mouse.move(
    dragElBoundingBox.x + dragElBoundingBox.width / 2,
    dragElBoundingBox.y + dragElBoundingBox.height / 4,
  )
  await page.mouse.down()
  await page.mouse.move(
    dragElBoundingBox.x + dragElBoundingBox.width / 2,
    dragElBoundingBox.y + dragElBoundingBox.height / 4 * 3,
    {
      steps: 50,
    },
  )
  await page.mouse.up()

  return { lastDragEl, previousLastDragElBoundingBox }
}

describe.concurrent('drag', async () => {
  it('events', async () => {
    const events = ['c.drag.start', 'c.drag.end']
    const page = await browser.newPage()
    await page.goto('http://127.0.0.1:4173/drag.html')

    await page.evaluate(() => {
      const el = document.querySelector('.c-drag')
      if (!el)
        return
      ['c.drag.start', 'c.drag.end'].forEach((e) => {
        el.addEventListener(e, () => {
          // eslint-disable-next-line no-console
          console.log(e)
        })
      })
    })

    const startEvent = page.waitForEvent('console', msg => msg.text().includes('c.drag.start'))
    const endEvent = page.waitForEvent('console', msg => msg.text().includes('c.drag.end'))

    await dragBox(page)

    const result = (await Promise.all([
      startEvent,
      endEvent,
    ])).map(item => item.text())

    events.forEach(e => expect(result).toContain(e))
  })

  // mouse drag doesn't work with playwright
  it.skip('mouse drag', async () => {
    const page = await browser.newPage()
    await page.goto('http://127.0.0.1:4173/drag.html')
    const { lastDragEl, previousLastDragElBoundingBox } = await dragBox(page)
    // await page.pause()

    if (!lastDragEl || !previousLastDragElBoundingBox)
      return

    const currentLastDragElBoundingBox = await lastDragEl.boundingBox()
    expect(currentLastDragElBoundingBox).not.toBeNull()
    expect(currentLastDragElBoundingBox).not.toEqual(previousLastDragElBoundingBox)
  })
})
