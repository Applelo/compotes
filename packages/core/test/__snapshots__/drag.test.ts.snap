// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`drag > destroy 1`] = `
"<!DOCTYPE html><html lang="en"><head>
    <link rel="icon" href="data:image/svg+xml,<svg xmlns=%22http://www.w3.org/2000/svg%22 viewBox=%220 0 100 100%22><text y=%22.9em%22 font-size=%2290%22>🍯</text></svg>">
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Drag</title>
    <script type="module" crossorigin="" src="/assets/drag-CCVgOnB8.js"></script>
    <link rel="modulepreload" crossorigin="" href="/assets/_parent-PLeNLLls.js">
    <link rel="stylesheet" crossorigin="" href="/assets/marquee-ybAlwBfl.css">
  </head>
  <body>
    <div class="c-drag" style="width: 200px; height: 200px;">
      <p>Lorem ipsum <a href="#">dolor sit amet</a> consectetur adipisicing elit. Consequatur, id? Dolorum cum est ipsum praesentium mollitia quod saepe nemo doloribus, aliquid dolore laudantium consequuntur alias facere veritatis magni, adipisci fugit.</p>
      <p>Lorem ipsum dolor sit, amet consectetur adipisicing elit. Pariatur, est dolore. Vel ducimus incidunt labore, ipsum necessitatibus omnis qui! Consectetur molestiae ipsam repellendus illum hic veritatis. <a href="#">Consequuntur impedit</a> animi voluptatum.</p>
    </div>
    <button class="js-destroy">
      Destroy
    </button>
  

</body></html>"
`;
